crypto/evp/keymgmt_lib.c

```diff
@@ -47,33 +47,60 @@ void *evp_keymgmt_util_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
 {
     void *keydata = NULL;
     struct import_data_st import_data;
-    size_t i, j;
+    size_t i = 0;
 
     /* Export to where? */
     if (keymgmt == NULL)
         return NULL;
 
     /* If we have an unassigned key, give up */
-    if (pk->pkeys[0].keymgmt == NULL)
+    if (pk->keymgmt == NULL)
         return NULL;
 
+    /* If |keymgmt| matches the "origin" |keymgmt|, no more to do */
+    if (pk->keymgmt == keymgmt)
```

Comment:
If we had a provider with an single alternate implementation of an algorithm I could potentially see it wanting to share the same key manager code.. The code makes the assumption currently that each provider has its own key manager.

Reply:
I don't think we should handle multiple implementations in a single provider, nor expect that such a mythical provider will have to externalize its key management between implementations.

At least not in this intial release.

Reply:
Multiple implementations in a single provider? Where did you get that idea from?

Reply:
Did you not read Shane's comment just before mine?

Reply:
Well, let's then say that we have a provider that implements SIGNATURE and KEYEXCH with some specific key. That means that it's reasonable to believe that they use the same key structure. Are you saying that each of them should implement their own handling of that same structure? Besides, what would an EVP_PKEY hang on to in such a scenario?

So yes, the current code not only assumes that every provider any asymmetric computation also comes with a keymgmt that holds on to the keys themselves, but *demands* it. 

This, BTW, has a corresponding thing in the legacy code: EVP_PKEY_ASN1_METHOD. There is one for every key type we support.

Reply:
> If we had a provider with an single alternate implementation of an algorithm I could potentially see it wanting to share the same key manager code..

So two providers where the OSSL_DISPATCH for the KEYMGMT happens to be the same, or even better, things like `->keymgmt-export` (to take an example)?  I've thought about the former, but haven't acted on it yet.  The latter gives me the heeby-jeebies for the moment, I need to investigate that line of thinking a little deeper.

Reply:
Either way, I think that's for another PR, not this.